Jetpack Social: Prevent SIG from enabling itself automatically on existing posts

danielpost opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
This PR fixes an issue where Social Image Generator was being enabled automatically for existing posts, by relying on Gutenberg to update the post meta rather than doing it a hook after saving the post. I've explained the changes in detail in the code since I think the context is helpful.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
1202816441505085-as-1204236160532606/f

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No.

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Make sure Social Image Generator is enabled for your site.
* Create a new post and make sure the Social Image Generator toggle is enabled.
* Publish it or save it as a draft.
* Check that the meta was saved correctly: `get_post_meta( POST_ID )` in the shell
* Now, delete the post meta (`delete_post_meta( POST_ID, '_wpas_options' )` and open the editor for the post.
* Check that the toggle is disabled.
* Change the post title or content and save it.
* Check the meta again, and make sure that SIG is now disabled.
* Repeat this process for the Classic Editor. In this case, SIG should always be disabled.

Commit:
[not verified] Prevent SIG from enabling itself automatically on existing posts

danielpost commented on projects/js-packages/publicize-components/src/components/social-image-generator/panel/index.js:
```diff
@@ -31,6 +31,10 @@ const SocialImageGeneratorPanel = ( { prePublish = false } ) => {
 		setTemplate,
 	} = useImageGeneratorConfig();
 
+	useEffect( () => {
+		setIsEnabled( isEnabled );
+	}, [] ); // eslint-disable-line react-hooks/exhaustive-deps
```
When creating a new post or editing an existing one, the [default value for enabling SIG is already set here](https://github.com/Automattic/jetpack/blob/trunk/projects/js-packages/publicize-components/src/hooks/use-image-generator-config/index.js#L8). However, this value was not being updated in the post meta. This `useEffect` hook simply updates the meta with the current value when the SIG panel is opened, so that it gets saved properly.

ESLint is being ignored because it wants `setIsEnabled` and `isEnabled` in the dependency array, which causes an infinite loop.

The tradeoff here is that SIG won't be enabled if a post is created programatically, since the post meta won't be present. This however can be fixed fairly easily by adding the post meta when the post is being created.

danielpost commented on projects/packages/publicize/src/class-publicize-setup.php:
```diff
@@ -44,6 +44,7 @@ public static function on_jetpack_feature_publicize_enabled() {
 		add_action( 'rest_api_init', array( new REST_Controller(), 'register_rest_routes' ) );
 		add_action( 'rest_api_init', array( new Social_Image_Generator\REST_Settings_Controller(), 'register_routes' ) );
 
+		self::init_social_image_generator();
```
**JUST FOR TESTING PURPOSES!**

I will remove this before merging because we will use https://github.com/Automattic/jetpack/pull/29529 instead.

danielpost commented on projects/packages/publicize/src/social-image-generator/class-setup.php:
```diff
@@ -88,17 +88,12 @@ public function set_meta( $post_id, $post ) {
 		}
 
-		// If we're here, it's safe to assume SIG should be enabled.
-		$post_settings->update_setting( 'enabled', true );
```
Since we now always have the meta present if SIG is enabled, we can simplify our check and remove the meta update in this hook.

Commit:
[not verified] Update meta on server instead

Commit:
[not verified] Improve setting check

pablinos committed:
[not verified] Set the default value for SIG enabled based on the post_status.

pablinos committed:
[not verified] Pass in the blog ID when checking if SIG is enabled.

We should be able to remove this when we merge the refactor of the init
code.

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1128,6 +1129,25 @@ public function register_post_meta() {
+		add_filter(
+			'default_post_metadata',
+			function ( $value, $object_id, $meta_key ) {
+				if ( $meta_key === '_wpas_options' ) {
```
This should probably be swapped to using the constant

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1128,6 +1129,25 @@ public function register_post_meta() {
+
+		add_filter(
+			'default_post_metadata',
+			function ( $value, $object_id, $meta_key ) {
```
It might be better if this is a method on the class, so it's easier for someone to remove, if needs be.

pablinos commented on projects/packages/publicize/src/class-publicize-base.php:
```diff
@@ -1128,6 +1129,25 @@ public function register_post_meta() {
+					if (
+						! empty( $post->post_status ) &&
+						'auto-draft' !== $post->post_status &&
+						! empty( $value[0]['image_generator_settings']['enabled'] )
```
I don't quite understand why the `value` array has a `0` index, like there would be multiple default values, but this does work.

The other thing to check is if a post has a value for `_wpas_options` (because they've shared an image for example) will we still check the defaults for `image_generator_settings`? If not, then we should probably flip the logic, so the default is `false` but we set it to `true` for `auto-draft` posts.

pablinos replied:
Ok, I've realised what's going on here now. When `get_metadata` is called with `$single=false` then the value is expected to be an array, and so we get passed the default value as an array of length 1.

I've reworked this function a bit to cope with both values for `$single` as, between the tests, and the default loading of a post in the block editor, it can be set differently.

pablinos committed:
[not verified] Remove unnecessary changelog.

pablinos committed:
[not verified] Swap to using the new feature check method.

pablinos committed:
[not verified] Refactor the method to filter the metadata.

pablinos committed:
[not verified] Prevent tokens from being generated when importing.

pablinos committed:
[not verified] Rework the default metadata hook to cope with $single.

Depending on whether the metadata is retrieved with `single` set, the
default metadata needs to get filtered differently.

This also broke the tests. So they had to be reworked too. It's thrown
up a bug with WorDBless as well, so I've had to skip a test.

pablinos committed:
[not verified] Remove the unnecessary call to tearDown.

pablinos committed:
[not verified] Update the test skipped comment.

gmjuhasz reviewed (dismissed):
This tests well for me, nice one!

pablinos committed:
Flip the logic, so SIG is disabled by default.

pablinos committed:
Update the SIG setting to enabled for new posts.

pablinos committed:
Explicitly update the metadata when it's a new, uneditted, post.

pablinos committed:
Reorder checks before generating the token.

pablinos committed:
Merge remote-tracking branch 'origin/trunk' into fix/jetpack-social-sig-enabling-automatically-on-existing-posts

pablinos committed:
Merge remote-tracking branch 'origin/trunk' into fix/jetpack-social-sig-enabling-automatically-on-existing-posts

pablinos committed:
Don't enable SIG unless it's enabled in settings.

danielpost commented:
I can't approve this because I opened the PR but this works as expected!

pablinos reviewed (approved):
Approving on behalf of @danielpost 😄
